### ISSUE TITLE
Yolov8s no bounding box on default settings

### ISSUE THREAD

**Issue by lida2003** [marcoslucianops/DeepStream-Yolo#597]
The issue is quite similar to https://github.com/marcoslucianops/DeepStream-Yolo/issues/390, but I need to use latest up to date versions.

Here is the video when I test yolov8s: https://drive.google.com/file/d/1I5MGC9_91h0drNASEM2z9VQUDptLNW_4/view?usp=drive_link

yolov4 is OK, https://drive.google.com/file/d/1bIdyqcfNa6JbuOyBR6NYOjnPqPTp-o-m/view?usp=sharing

```
Software part of jetson-stats 4.2.12 - (c) 2024, Raffaello Bonghi
Model: NVIDIA Orin Nano Developer Kit - Jetpack 5.1.4 [L4T 35.6.0]
NV Power Mode[0]: 15W
Serial Number: [XXX Show with: jetson_release -s XXX]
Hardware:
 - P-Number: p3767-0005
 - Module: NVIDIA Jetson Orin Nano (Developer kit)
Platform:
 - Distribution: Ubuntu 20.04 focal
 - Release: 5.10.216-tegra
jtop:
 - Version: 4.2.12
 - Service: Active
Libraries:
 - CUDA: 11.4.315
 - cuDNN: 8.6.0.166
 - TensorRT: 8.5.2.2
 - VPI: 2.4.8
 - OpenCV: 4.9.0 - with CUDA: YES
DeepStream C/C++ SDK version: 6.3

Python Environment:
Python 3.8.10
    GStreamer:                   YES (1.16.3)
  NVIDIA CUDA:                   YES (ver 11.4, CUFFT CUBLAS FAST_MATH)
        OpenCV version: 4.9.0  CUDA True
          YOLO version: 8.3.33
         Torch version: 2.1.0a0+41361538.nv23.06
   Torchvision version: 0.16.1+fdea156
DeepStream SDK version: 1.1.8
```

**Comment by PaoXi** [comment]
Did you solve this error, the default yolov8s after running it with deepstream-app -c deep* no detection is shown no bounding boxes

**Comment by lida2003** [comment]
@PaoXi No, I didn't have time to dig into it. And I didn't find any clue yet.

**Comment by marcoslucianops** [comment]
@PaoXi is your board also Orin Nano?

**Comment by PaoXi** [comment]
Yes it's 16 Gb

**Comment by marcoslucianops** [comment]
Can someone send me the exported onnx file (from Orin Nano)?

**Comment by PaoXi** [comment]
There're two version, one exported using ultralytics API. other one directly with pytorch script 

https://www.mediafire.com/file/edzascweikxrup9/yolov8s.pt.onnx/file
https://www.mediafire.com/file/5qxn3sxbqhym53q/yolov8sss.onnx/file

**Comment by PaoXi** [comment]
They're two version, one with ultralytics's API, and the other using directly pytorch

https://www.mediafire.com/file/edzascweikxrup9/yolov8s.pt.onnx/file
https://www.mediafire.com/file/5qxn3sxbqhym53q/yolov8sss.onnx/file

**Comment by lida2003** [comment]
@marcoslucianops I don't know exact model version, but I have downloaded [here](https://docs.ultralytics.com/models/yolov8/#supported-tasks-and-modes).

export from pt to onnx command:

```bash
$ yolo export model=yolov8s.pt format=onnx
$ yolo version
8.3.33
```

Attached below:
- [yolov8s.pt](https://drive.google.com/file/d/1cZjx_98fI-c2fCg3Mf_A7ED0NlcCq6dF/view?usp=sharing)
- [yolov8s.onnx](https://drive.google.com/file/d/1NSPHianqfCc685mwAp02b_NKuwq6F-rF/view?usp=sharing)

**Comment by marcoslucianops** [comment]
@lida2003 you need to export with the `utils/export_yoloV8.py`. See the `docs/YOLOv8.md`.

**Comment by lida2003** [comment]
@marcoslucianops The result is the same (no bounding boxes), but I got the right way to generate onnx file, thanks. 

PS: delete `model_b1_gpu0_fp32.engine` before run `deepstream -c deep*`.

- [yolov8s.pt.onnx](https://drive.google.com/file/d/1PqJ0lYyLROLyWYDi-q5KzKUzITTFn_IP/view?usp=sharing)

**Comment by lida2003** [comment]
@marcoslucianops We are trying to reproduce the performance mentioned in the following [link](https://github.com/ultralytics/ultralytics/issues/17640#issuecomment-2537927750), which claims to achieve 181 FPS with INT8 precision on Jetson Orin NX. However, we are currently stuck on the bounding box selection issue. Any good suggestions?

**Comment by marcoslucianops** [comment]
I didn't get detection with your models @lida2003 and @PaoXi, but I got detections converting here. What is the PyTorch version you are using?

**Comment by marcoslucianops** [comment]
Can you try `v2.0.0` and `v1.14.0`?

**Comment by lida2003** [comment]
> Can you try `v2.0.0` and `v1.14.0`?

No, on this board runing jetpack 5.1.3/5.1.4 there is only one release version [torch-2.1.0a0+41361538.nv23.06-cp38-cp38-linux_aarch64.whl](https://developer.download.nvidia.cn/compute/redist/jp/v512/pytorch/torch-2.1.0a0+41361538.nv23.06-cp38-cp38-linux_aarch64.whl), which is in [developer.download.nvidia.cn/compute/redist/jp/v512/pytorch/](https://developer.download.nvidia.cn/compute/redist/jp/v512/pytorch/). Oler versions might be removed links from their web server.

**I will try to export in x86 env and to see if it helps, and get back to you soon.**

Also I have made a request here: [[REQUEST] build script for pytorch or up to date pytorh binary release supporting jetson boards running L4T35.6(ubuntu20.04)](https://forums.developer.nvidia.com/t/request-build-script-for-pytorch-or-up-to-date-pytorh-binary-release-supporting-jetson-boards-running-l4t35-6-ubuntu20-04/316972)

**Comment by marcoslucianops** [comment]
I think the versions I said also works on 5.1.3/5.1.4. Can you try?

**Comment by lida2003** [comment]
@marcoslucianops Do you mean it might be related with PyTorch version?

- [NG] As I'm using 2.1 (from nvidia), I'm NOT sure what version @PaoXi was using.
- [OK] From previous talk, you were using 2.4 (Not sure GPU or CPU version).

We don't have other Pytorch GPU version for L4T35.6 except 2.1. So only possible version is CPU versions which might be v2.0.0 or v1.14.0, which you are refering.

**Comment by marcoslucianops** [comment]
Please try:
- `v2.0.0`: https://nvidia.box.com/shared/static/i8pukc49h3lhak4kkn67tg9j4goqm0m7.whl
- `v1.14.0`: https://developer.download.nvidia.com/compute/redist/jp/v51/pytorch/torch-1.14.0a0+44dac51c.nv23.02-cp38-cp38-linux_aarch64.whl

**Comment by lida2003** [comment]
> @marcoslucianops Do you mean it might be related with PyTorch version?

Well, don't have much time today. But ... ... **It's great! It works using below exported onnx file to my jetson orin board.**

```
          YOLO version: 8.3.50
         Torch version: 2.5.1+cu124
   Torchvision version: 0.20.1+cu124
```

```bash
$ python3 ./utils/export_yoloV8.py -w yolov8s.pt --dynamic
```

- [yolov8s.pt.x86.onnx](https://drive.google.com/file/d/1ae3fIAv-2eEdXNX9yEu7r6_4aOAUAsnj/view?usp=sharing) works pretty fine. And the speed is OK, stable at 30FPS (file:///opt/nvidia/deepstream/deepstream/samples/streams/sample_1080p_h265.mp4)

Note1: `model_b1_gpu0_fp32.engine` removed before using new onnx file.
Note2: Well, if needed I will try `yolo` command/v2.0.0/v1.14.0 if I have time and get my current job done.

_PS: The above test takes quite a lot of time for downloading python components._

**Comment by PaoXi** [comment]
Thanks, @lida2003 for your comment and for sharing this approach! This version works pretty fine for me as well. I'm grateful for your insights—it's been super helpful.

By the way, could you share more details about how you exported this ONNX version? I'm curious if there are specific steps or tweaks you used that made it work so well.

**Comment by lida2003** [comment]
> By the way, could you share more details about how you exported this ONNX version? I'm curious if there are specific steps or tweaks you used that made it work so well.

No special steps, just use 2.5.1 on x86 to export the onnx as guide said.

But I have found some issue related with BYTETrack here: https://github.com/marcoslucianops/DeepStream-Yolo/issues/605. Not sure if it's related with onnx file.


> @marcoslucianops Do you mean it might be related with PyTorch version?
> 
>     * [NG] As I'm using 2.1 (from nvidia), I'm NOT sure what version @PaoXi was using.
> 
>     * [OK] From previous talk, you were using 2.4 (Not sure GPU or CPU version).

- [OK] 2.5.1+cu124, x86 exported onnx file
- [OK] 1.14.0a0+44dac51c.nv23.2 , jetson Orin NX exported onnx file

@marcoslucianops we have the above result. I'm trying build jetson orin pytorch 2.5.1 still some issue now.

**Comment by PaoXi** [comment]
Thanks for sharing, @lida2003. It’s great to know the export worked with 2.5.1 on x86 arch.

@marcoslucianops, regarding a custom model with 6 classes instead of 80, I believe we’d need to customize the parsing function to handle the changes. Could you confirm if this is the right approach or if there are additional steps needed for the modification?

Thanks

**Comment by lida2003** [comment]
@sdhamodaran It might be something to do with onnxruntime. Check onnxruntime version into account also.

**Comment by sdhamodaran** [comment]
Thanks for your reply. I have '1.19.2' version of onnxruntime. Which one you have?

**Comment by lida2003** [comment]
@sdhamodaran I use onnxruntime_gpu-1.17.0-cp38-cp38-linux_aarch64.whl, which comes from nvidia, on jetson orin.

I have upgraded pytorch to [pytorch-v2.5.1+l4t35.6-cp38-cp38-aarch64](https://github.com/SnapDragonfly/pytorch/releases/tag/v2.5.1%2Bl4t35.6-cp38-cp38-aarch64) on jetson orin L4T 35.6 Jetpack 5.1.4. And it seems still got trouble.

Right now, I'm compiling onnxruntime 1.19.2, still got issue on jetson orin L4T 35.6 Jetpack 5.1.4.

**Comment by PaoXi** [comment]
Have you tried the last exported file by @lida2003 ? Try exporting the ONNX file on an x86 architecture workstation using the latest versions of CUDA, PyTorch, and ONNX Runtime. Then, use the generated ONNX file to export the engine file.

**Comment by sdhamodaran** [comment]
Hi @PaoXi  and @lida2003 , thanks for offering to use your exported file. That worked well. So it is the onnx file creation process that is causing the issue.

**Comment by sdhamodaran** [comment]
Also a general question. When you guys try different combos of torch, torchvision and onnxruntime, from which step do you repeat the process from. I am starting from creation of onnx file and then running deepstream-app command. Am i doing it right?

**Comment by lida2003** [comment]
> So it is the onnx file creation process that is causing the issue.

YES, but I didn't know the rootcause. Probably onnxruntime, Need to confirm.

When x86 onnx file works, i thought it was pytorch version issue. After ugraded to 2.5.1 on jetson, I found it will still fail exporting on jetson, so I think it might be onnxruntime now.

Anyway, x86's onnx file works. We need more tests or more experienced person to take time and effort to find out rootcause.

>  I am starting from creation of onnx file and then running deepstream-app command. Am i doing it right?

YES, you are absolutely right.

**Comment by lida2003** [comment]
@marcoslucianops As you can see in my first post of this thread, I didn't noticed onnxruntime. But I checked my system, there are two onnxruntime w/o gpu. How can I distinguish whether the application is actually using `onnxruntime` or `onnxruntime-gpu` in the code?

`python3 ./utils/export_yoloV8.py -w yolov8s.pt --dynamic`

```
$ sudo ./wrapper.sh version
Skipping CMD_KEYMONITOR execution for module 'version'.
Executing command on module version:

Software part of jetson-stats 4.2.12 - (c) 2024, Raffaello Bonghi
Model: NVIDIA Orin Nano Developer Kit - Jetpack 5.1.4 [L4T 35.6.0]
NV Power Mode[0]: 15W
Serial Number: [XXX Show with: jetson_release -s XXX]
Hardware:
 - P-Number: p3767-0005
 - Module: NVIDIA Jetson Orin Nano (Developer kit)
Platform:
 - Distribution: Ubuntu 20.04 focal
 - Release: 5.10.216-tegra
jtop:
 - Version: 4.2.12
 - Service: Active
Libraries:
 - CUDA: 11.8.89
 - cuDNN: 8.6.0.166
 - TensorRT: 8.5.2.2
 - VPI: 2.4.8
 - Vulkan: 1.3.204
 - OpenCV: 4.9.0 - with CUDA: YES
DeepStream C/C++ SDK version: 6.3

Python Environment:
Python 3.8.10
    GStreamer:                   YES (1.16.3)
  NVIDIA CUDA:                   YES (ver 11.4, CUFFT CUBLAS FAST_MATH)
         OpenCV version: 4.9.0  CUDA True
           YOLO version: 8.3.33
         PYCUDA version: 2024.1.2
          Torch version: 2.5.1+l4t35.6
    Torchvision version: 0.20.1a0+3ac97aa
 DeepStream SDK version: 1.1.8
onnxruntime     version: 1.16.3
onnxruntime-gpu version: 1.18.0
```

**Comment by marcoslucianops** [comment]
The issue is related to the PyTorch in Jetson.

**Comment by NiteshMin** [comment]
@marcoslucianops which pytorch version do you recommend for Jetpack 5.1.3 as i am having the same issue of jetson xavier ,orin nano.

**Comment by zqz85** [comment]
这个问题的原因是pytorch引起的。torch-2.1.2-cp38-cp38-manylinux2014_aarch64.whl你必须使用这个版本才能正常显示。而jetson所推荐的pytorch对于jetpack5.1.4版本支持并不好，包括一些网友提供的2.5.1版本也不行。

**Comment by zqz85** [comment]
> > 这个问题的原因是pytorch引起的。torch-2.1.2-cp38-cp38-manylinux2014_aarch64.whl你必须使用这个版本才能正常显示。而jetson所推荐的pytorch对于jetpack5.1.4版本支持并不好，包括一些网友提供的2.5.1版本也不行。
> 
> Hi在哪里下载此whl文件：torch-2.1.2-cp38-cp38-manylinux2014_aarch64.whl [@zqz85](https://github.com/zqz85)
> 
> 我找到了，https://forums.developer.nvidia.com/t/pytorch-for-jetson/72048
我想要上传whl文件，但是系统并不允许。github好像不让上传whl相关文件。你可以发一封邮件到hanwujizhuzhu@163.com，如果你还没找到的话。

**Comment by marcoslucianops** [comment]
@NiteshMin I don't which version would work. Try the suggestion from @zqz85 or this version

`pip3 install https://developer.download.nvidia.cn/compute/redist/jp/v512/pytorch/torch-2.1.0a0+41361538.nv23.06-cp38-cp38-linux_aarch64.whl`
